codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/EndpointParamsKnowledgeIndex.java

```diff
@@ -0,0 +1,190 @@
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.codegen.poet.rules;
+
+import static javax.lang.model.element.Modifier.PRIVATE;
+import static javax.lang.model.element.Modifier.STATIC;
+
+import com.squareup.javapoet.ClassName;
+import com.squareup.javapoet.MethodSpec;
+import com.squareup.javapoet.ParameterizedTypeName;
+import com.squareup.javapoet.TypeName;
+import com.squareup.javapoet.TypeVariableName;
+import java.util.EnumMap;
+import java.util.Map;
+import java.util.Optional;
+import javax.lang.model.element.Modifier;
+import software.amazon.awssdk.awscore.client.config.AwsClientOption;
+import software.amazon.awssdk.awscore.endpoints.AccountIdEndpointMode;
+import software.amazon.awssdk.awscore.endpoints.AccountIdEndpointModeResolver;
+import software.amazon.awssdk.codegen.internal.Utils;
+import software.amazon.awssdk.codegen.model.intermediate.IntermediateModel;
+import software.amazon.awssdk.codegen.model.internal.LocalParameter;
+import software.amazon.awssdk.codegen.model.rules.endpoints.BuiltInParameter;
+import software.amazon.awssdk.codegen.model.rules.endpoints.ParameterModel;
+import software.amazon.awssdk.codegen.poet.PoetUtils;
+import software.amazon.awssdk.core.SelectedAuthScheme;
+import software.amazon.awssdk.core.client.config.SdkClientConfiguration;
+import software.amazon.awssdk.core.client.config.SdkClientOption;
+import software.amazon.awssdk.identity.spi.AwsCredentialsIdentity;
+import software.amazon.awssdk.identity.spi.Identity;
+import software.amazon.awssdk.utils.CompletableFutureUtils;
+import software.amazon.awssdk.utils.internal.CodegenNamingUtils;
+
+/**
+ * Knowledge index to get access to endpoint parameters known to the client builder classes.
+ */
+public final class EndpointParamsKnowledgeIndex {
+    private static final Map<BuiltInParameter, LocalParameter> BUILT_IN_PARAMS_FOR_CLIENT_BUILDER =
+        new EnumMap<>(BuiltInParameter.class);
+    private final IntermediateModel intermediateModel;
+    private Map<BuiltInParameter, LocalParameter> parametersToGenerate = new EnumMap<>(BuiltInParameter.class);
+
+    static {
+        BUILT_IN_PARAMS_FOR_CLIENT_BUILDER.put(
+            BuiltInParameter.AWS_AUTH_ACCOUNT_ID_ENDPOINT_MODE,
+            new LocalParameter("accountIdEndpointMode",
+                               AccountIdEndpointMode.class,
+                                     "Sets the behavior when account ID based endpoints are created. "
+                                     + "See {@link AccountIdEndpointMode} for values"));
```

Comment:
minor: when referencing classes in a javadoc, we should still use `$T` placeholder in case the class is not already imported by something else

Reply:
Thanks for the reminder, will fix!